Telegram/SourceFiles/platform/linux/linux_xdp_open_with_dialog.cpp

```diff
@@ -0,0 +1,180 @@
+/*
+This file is part of Telegram Desktop,
+the official desktop application for the Telegram messaging service.
+
+For license and copyright information please follow this link:
+https://github.com/telegramdesktop/tdesktop/blob/master/LEGAL
+*/
+#include "platform/linux/linux_xdp_open_with_dialog.h"
+
+#include "base/platform/base_platform_info.h"
+#include "base/platform/linux/base_linux_glibmm_helper.h"
+#include "core/application.h"
+#include "window/window_controller.h"
+#include "base/openssl_help.h"
+
+#include <QtGui/QWindow>
+
+#include <fcntl.h>
+#include <gio/gunixfdlist.h>
+#include <glibmm.h>
+#include <giomm.h>
+#include <private/qguiapplication_p.h>
+
+namespace Platform {
+namespace File {
+namespace internal {
+namespace {
+
+constexpr auto kXDGDesktopPortalService = "org.freedesktop.portal.Desktop"_cs;
+constexpr auto kXDGDesktopPortalObjectPath = "/org/freedesktop/portal/desktop"_cs;
+constexpr auto kXDGDesktopPortalOpenURIInterface = "org.freedesktop.portal.OpenURI"_cs;
+constexpr auto kPropertiesInterface = "org.freedesktop.DBus.Properties"_cs;
+
+class XDPOpenWithDialog : public QWindow {
+public:
+	XDPOpenWithDialog(const QString &filepath)
+	: _filepath(filepath.toStdString()) {
+	}
+
+	bool exec();
+
+private:
+	Glib::ustring _filepath;
+};
+
+bool XDPOpenWithDialog::exec() {
+	try {
+		const auto connection = Gio::DBus::Connection::get_sync(
+			Gio::DBus::BusType::BUS_TYPE_SESSION);
+
+		auto reply = connection->call_sync(
+			std::string(kXDGDesktopPortalObjectPath),
+			std::string(kPropertiesInterface),
+			"Get",
+			base::Platform::MakeGlibVariant(std::tuple{
+				Glib::ustring(
+					std::string(kXDGDesktopPortalOpenURIInterface)),
+				Glib::ustring("version"),
+			}),
+			std::string(kXDGDesktopPortalService));
+
+		const auto version = base::Platform::GlibVariantCast<uint>(
+			base::Platform::GlibVariantCast<Glib::VariantBase>(
+				reply.get_child(0)));
+
+		if (version < 3) {
+			return false;
+		}
+
+		const auto fd = open(
+			_filepath.c_str(),
+			O_RDONLY);
+
+		if (fd == -1) {
+			return false;
+		}
+
+		const auto fdGuard = gsl::finally([&] { ::close(fd); });
+		auto outFdList = Glib::RefPtr<Gio::UnixFDList>();
+
+		const auto parentWindowId = [&]() -> Glib::ustring {
+			std::stringstream result;
+			if (const auto activeWindow = Core::App().activeWindow()) {
+				if (IsX11()) {
```

Comment:
In C++17 you can even join this `if` condition and the one above.
```suggestion
			if (const auto activeWindow = Core::App().activeWindow(); IsX11()) {
```